### PR TITLE
Add adUDotDot and infrastructure

### LINKED ISSUE #18178: Add second time derivative support for AD variables
## Reason
For AD objects that require second time derivatives, an `adUDotDot` option doesn't exist. This should be added.

## Design
Add `adUDotDot` and `adUDotDotNeighbor` in `MooseVariable*` classes, and implement in the Newmark-Beta and Central Difference time integrators. Add an `adCoupledDotDot` method for developers to use in custom object param definitions.

## Impact
New capability.

### PR DESCRIPTION
**Pull request by cticenhour**
Closes #18178

### CODE REVIEW

**Comment by cticenhour** [comment]
FYI - GitHub testing status did not update properly. All tests are green: https://civet.inl.gov/event/70202/

**Comment by cticenhour** [comment]
@rwcarlsen Could you give this a review when you get a chance? Apologies for the potential spam - it doesn't have you listed as a "participant" yet, even though it tagged you as a Code Owner.

File: framework/include/interfaces/Coupleable.h
Comment (lindsayad) [inline framework/include/interfaces/Coupleable.h:780]:
```suggestion
   * @return Reference to an ADVariableValue containing the second time derivative of the coupled
```

File: framework/include/variables/MooseVariableData.h
Comment (lindsayad) [inline framework/include/variables/MooseVariableData.h:330]: I've been a guilty culprit in the past, but for things like this I'd prefer to keep the class definition clean of multi-line method definitions. You can put this outside the class definition as an `inline` function in the header

File: framework/include/variables/MooseVariableDataFV.h
Comment (lindsayad) [inline framework/include/variables/MooseVariableDataFV.h:196]: same as above

File: framework/src/interfaces/Coupleable.C
Comment (lindsayad) [inline framework/src/interfaces/Coupleable.C:1782]:
```suggestion
  const auto * const var = getVarHelper<MooseVariableField<Real>>(var_name, comp);
```

File: test/tests/time_integrators/central-difference/ad_central_difference_dotdot.i
Comment (lindsayad) [inline test/tests/time_integrators/central-difference/ad_central_difference_dotdot.i:9]: This is fancy. What does this mean?

File: test/tests/time_integrators/central-difference/tests
Comment (lindsayad) [inline test/tests/time_integrators/central-difference/tests:18]: We should try to keep our requirements as implementation agnostic as possible, so probably something more like
```suggestion
    requirement = "A central differencing time integrator shall correctly compute the second"
                  " time derivative of an automatic differentiation variable."
```

File: test/tests/time_integrators/newmark-beta/tests
Comment (lindsayad) [inline test/tests/time_integrators/newmark-beta/tests:30]:
```suggestion
    requirement = "A Newmark-beta method time integrator shall correctly compute the second"
                  " time derivative of an automatic differentiation variable using the default beta"
                  " and gamma parameters."
```

**Review by lindsayad — DISMISSED** [review]
Overall looks good to me

**Review by rwcarlsen — DISMISSED** [review]

File: test/tests/time_integrators/central-difference/ad_central_difference_dotdot.i
Comment (cticenhour) [reply test/tests/time_integrators/central-difference/ad_central_difference_dotdot.i:9]: I am not entirely sure....precursor to the new SQA requirements I think, since this is copypasta from an old test. I will remove in the same commit that I switch to `inline` functions for `adUDotDot`

File: framework/include/variables/MooseVariableData.h
Comment (cticenhour) [reply framework/include/variables/MooseVariableData.h:330]: Would you like me to move `adUDot` as well? That was here before I made these changes.

Comment (lindsayad) [reply framework/include/variables/MooseVariableData.h:330]: You don't need to, but if you're pushing any more commits, feel free to

Comment (cticenhour) [reply framework/include/variables/MooseVariableData.h:330]: Will do

Comment (cticenhour) [reply framework/include/variables/MooseVariableData.h]: Do I need to be explicit about adding the `inline` to a method that isn't fully specialized? Take a look at 2fdf84b. The comments already present in `MooseVariableData.h` at the bottom of the file made that a little unclear.

File: test/tests/time_integrators/central-difference/ad_central_difference_dotdot.i
Comment (cticenhour) [reply test/tests/time_integrators/central-difference/ad_central_difference_dotdot.i:9]: Resolved in 2fdf84b

File: framework/include/variables/MooseVariableData.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableData.h]: No you do not. I forgot that this was a templated class when I made my original comment. Any thing with a template parameter will be automatically `inline`

**Comment by cticenhour** [comment]
@rwcarlsen I've defined the `adUDotDot` method for FV, as you requested. A new test for it is in 3690605, but I'm having an issue with it. I modeled this test after the new `ad_newmark_beta_dotdot.i` test I created in fa7291b, but it doesn't seem to work correctly. See the output below - the first three timesteps achieve the expected output, but then I cannot duplicate the rest of the series with pen-and-paper calcs using the variable value and dot. Is there an issue with the Postprocessor I'm using that produces the element average second time derivative? Not being well-versed on FV, that was my first thought. 

```
+----------------+----------------+
| time           | udotdot        |
+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |
|   1.000000e-01 |   0.000000e+00 |
|   2.000000e-01 |   4.761905e-02 |
|   3.000000e-01 |   9.070295e-02 |
|   4.000000e-01 |   3.444552e-02 |
|   5.000000e-01 |  -1.645405e-02 |
|   6.000000e-01 |  -1.488700e-02 |
+----------------+----------------+
```

**Comment by rwcarlsen** [comment]
You could check the actual values from _udotdot[_qp] and see if they show what we expect (i.e. it reports the same value that the element average postprocessor is computing for its whole-elem integral).  For FV, we should be integrating over only a single qp - or at least - all qps should have copies of the single elemental value for an element.

**Comment by cticenhour** [comment]
Alright, I've traced things back to the actual time-integrator code, and convinced myself it was working properly. I decided my test was a bit shoddy (based on the newmark AD test I also added here), so I replaced it with an FV simple diffusion transient example which works a lot better. I think this is ready to a final review.

**Comment by cticenhour** [comment]
@lindsayad I also rebased this on top of your recent CNSFV changes due to conflicts, so double-check `MooseVariableDataFV` for anything I might have missed in the consolidation.

**Review by lindsayad — DISMISSED** [review]
LGTM

**Review by rwcarlsen — DISMISSED** [review]

**Comment by cticenhour** [comment]
Had to rebase again due to a conflict created yesterday while the test boxes were down. Could I get another approval?

**Review by rwcarlsen — APPROVED** [review]

**Pull request merged**
